File: src/dag/store.rs
use super::read::OwnedRead;
use super::write::Write;
use super::Result;
use crate::kv;
use crate::util::rlog::LogContext;

pub struct Store {
    kv: Box<dyn kv::Store>,
}

impl Store {
    pub fn new(kv: Box<dyn kv::Store>) -> Store {
        Store { kv }
    }

    pub async fn read(&self, lc: LogContext) -> Result<OwnedRead<'_>> {
        Ok(OwnedRead::new(self.kv.read(lc).await?))
    }

    pub async fn write(&self, lc: LogContext) -> Result<Write<'_>> {
        Ok(Write::new(self.kv.write(lc).await?))
    }

    pub async fn close(&self) {
        self.kv.close().await;
    }
}

File: src/dag/store.ts
import type {Chunk} from './chunk';
import type {ReadonlyJSONValue} from '../mod';
import type {Hash} from '../hash';

export interface Store {
  read(): Promise<Read>;
  withRead<R>(fn: (read: Read) => R | Promise<R>): Promise<R>;
  write(): Promise<Write>;
  withWrite<R>(fn: (Write: Write) => R | Promise<R>): Promise<R>;
  close(): Promise<void>;
}

interface GetChunk {
  getChunk(hash: Hash): Promise<Chunk | undefined>;
}

export interface Read extends GetChunk {
  hasChunk(hash: Hash): Promise<boolean>;
  mustGetChunk(hash: Hash): Promise<Chunk>;
  getHead(name: string): Promise<Hash | undefined>;
  close(): void;
  get closed(): boolean;
}

export interface Write extends Read {
  createChunk<V extends ReadonlyJSONValue>(
    data: V,
    refs: readonly Hash[],
  ): Chunk<V>;
  putChunk(c: Chunk): Promise<void>;
  setHead(name: string, hash: Hash): Promise<void>;
  removeHead(name: string): Promise<void>;
  assertValidHash(hash: Hash): void;
  commit(): Promise<void>;
}

export class ChunkNotFoundError extends Error {
  name = 'ChunkNotFoundError';
  readonly hash: Hash;
  constructor(hash: Hash) {
    super(`Chunk not found ${hash}`);
    this.hash = hash;
  }
}

export async function mustGetChunk(
  store: GetChunk,
  hash: Hash,
): Promise<Chunk> {
  const chunk = await store.getChunk(hash);
  if (chunk) {
    return chunk;
  }
  throw new ChunkNotFoundError(hash);
}
